plutus-core/plutus-core/src/Crypto/BLS12_381/Pairing.hs

```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications      #-}
+module Crypto.BLS12_381.Pairing
+    (
+     MlResult (..),
+     millerLoop,
+     mulMlResult,
+     finalVerify,
+     mlResultMemSizeBytes
+    ) where
+
+import Crypto.External.EllipticCurve.BLS12_381 qualified as BlstBindings
+import Crypto.External.EllipticCurve.BLS12_381.Internal qualified as BlstBindings.Internal
+
+import Crypto.BLS12_381.G1 qualified as G1
+import Crypto.BLS12_381.G2 qualified as G2
+import PlutusCore.Pretty.PrettyConst (ConstConfig)
+import Text.PrettyBy (PrettyBy, prettyBy)
+
+import Control.DeepSeq (NFData, rnf)
+import Data.Bifunctor (second)
+import Flat
+import Prettyprinter
+
+{- | This type represents the result of computing a pairing using the Miller
+   loop.  Values of this type are ephemeral, only created during script
+   execution.  We do not provide any means of serialising, deserialising,
+   printing, or parsing MlResult values. -}
+newtype MlResult = MlResult { unMlResult :: BlstBindings.PT }
+    deriving newtype (Eq)
+instance Show MlResult where
```

Comment:
> It would break all our stuff that requires all builtin types to be showable.
I think that's right.  I'll check.

[Time passes...]

It gets a surprisingly long way without that (including all the way through plutus-core and plutus-ir), but fails in some Hedgehog tests [here](https://github.com/input-output-hk/plutus/blob/411cf29bfbffbab6a4e6a31ef783cd52362affdf/plutus-core/testlib/PlutusCore/Generators/Hedgehog/TypeEvalCheck.hs#L118) and also on line 133 of the same file.  I tried putting a random string in there but then it failed in the NEAT tests.  I gave up at that point.